### ISSUE TITLE
Subscription overloads to allow weak referencing of targets

### ISSUE THREAD

**Issue by IanKeen** [RxSwiftCommunity/RxSwift-Ext#14]
### Subscription overloads to allow weak referencing of targets

Currently when you subscribe to an observable the pattern is something like:
```swift
let obs: Observable<String>!
obs.subscribeNext { [weak self] value in
    self.handleString(value)
}
```
This allows us to avoid retain cycles on `self` but is cumbersome, it is tempting to write the following:
```swift
let obs: Observable<String>!
obs.subscribeNext(self.handleString)
```
however we have now introduced a retain cycle on `self

### Motivation for inclusion
The second format is much more desirable but is obviously flawed - I'd like to introduce some overloads for subscription that get close to this format without the cycle...

### Example of use
Here is the full code for the extensions:
```swift
extension ObservableType {
    /**
     Leverages instance method currying to provide a weak wrapper around an instance function
     
     - parameter obj:    The object that owns the function
     - parameter method: The instance function represented as `InstanceType.instanceFunc`
     */
    private func weakify<A: AnyObject, B>(obj: A, method: ((A) -> (B) -> Void)?) -> ((B) -> Void) {
        return { [weak obj] value in
            guard let obj = obj else { return }
            method?(obj)(value)
        }
    }
    
    /**
     Subscribes an event handler to an observable sequence.
     
     - parameter weak: Weakly referenced object containing the target function.
     - parameter on: Function to invoke on `weak` for each event in the observable sequence.
     - returns: Subscription object used to unsubscribe from the observable sequence.
     */
    @warn_unused_result(message="http://git.io/rxs.ud")
    public func subscribe<A: AnyObject>(weak obj: A, _ on: (A) -> (RxSwift.Event<Self.E>) -> Void) -> Disposable {
        return self.subscribe(weakify(obj, method: on))
    }
    
    /**
     Subscribes an element handler, an error handler, a completion handler and disposed handler to an observable sequence.
     
     - parameter weak: Weakly referenced object containing the target function.
     - parameter onNext: Function to invoke on `weak` for each element in the observable sequence.
     - parameter onError: Function to invoke on `weak` upon errored termination of the observable sequence.
     - parameter onCompleted: Function to invoke on `weak` upon graceful termination of the observable sequence.
     - parameter onDisposed: Function to invoke on `weak` upon any type of termination of sequence (if the sequence has
     gracefully completed, errored, or if the generation is cancelled by disposing subscription)
     - returns: Subscription object used to unsubscribe from the observable sequence.
     */
    @warn_unused_result(message="http://git.io/rxs.ud")
    public func subscribe<A: AnyObject>(
        weak obj: A,
        onNext: ((A) -> (Self.E) -> Void)? = nil,
        onError: ((A) -> (ErrorType) -> Void)? = nil,
        onCompleted: ((A) -> () -> Void)? = nil,
        onDisposed: ((A) -> () -> Void)? = nil) -> Disposable {
            
            let disposable: Disposable
            
            if let disposed = onDisposed {
                disposable = AnonymousDisposable(weakify(obj, method: disposed))
            }
            else {
                disposable = NopDisposable.instance
            }
            
            
            let observer = AnyObserver { (e: RxSwift.Event<Self.E>) in
                switch e {
                case .Next(let value):
                    self.weakify(obj, method: onNext)(value)
                case .Error(let e):
                    self.weakify(obj, method: onError)(e)
                    disposable.dispose()
                case .Completed:
                    self.weakify(obj, method: onCompleted)()
                    disposable.dispose()
                }
            }
            
            return StableCompositeDisposable.create(
                self.asObservable().subscribe(observer),
                disposable
            )
    }
    
    /**
     Subscribes an element handler to an observable sequence.
     
     - parameter weak: Weakly referenced object containing the target function.
     - parameter onNext: Function to invoke on `weak` for each element in the observable sequence.
     - returns: Subscription object used to unsubscribe from the observable sequence.
     */
    @warn_unused_result(message="http://git.io/rxs.ud")
    public func subscribeNext<A: AnyObject>(weak obj: A, _ onNext: (A) -> (Self.E) -> Void) -> Disposable {
        return self.subscribeNext(weakify(obj, method: onNext))
    }
    
    /**
     Subscribes an error handler to an observable sequence.
     
     - parameter weak: Weakly referenced object containing the target function.
     - parameter onError: Function to invoke on `weak` upon errored termination of the observable sequence.
     - returns: Subscription object used to unsubscribe from the observable sequence.
     */
    @warn_unused_result(message="http://git.io/rxs.ud")
    public func subscribeError<A: AnyObject>(weak obj: A, _ onError: (A) -> (ErrorType) -> Void) -> Disposable {
        return self.subscribeError(weakify(obj, method: onError))
    }
    
    /**
     Subscribes a completion handler to an observable sequence.
     
     - parameter weak: Weakly referenced object containing the target function.
     - parameter onCompleted: Function to invoke on `weak` graceful termination of the observable sequence.
     - returns: Subscription object used to unsubscribe from the observable sequence.
     */
    @warn_unused_result(message="http://git.io/rxs.ud")
    public func subscribeCompleted<A: AnyObject>(weak obj: A, _ onCompleted: (A) -> () -> Void) -> Disposable {
        return self.subscribeCompleted(weakify(obj, method: onCompleted))
    }
}
```

And the sample usage:
```swift
class FooViewController: UIViewController {
    var obs: Observable<String>!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        obs.subscribeNext(weak: self, FooViewController.handleString)
    }
    
    private func handleString(string: String) {
        //..
    }
}
```

**Comment by IanKeen** [comment]
If approved, I would of course be happy to submit a PR :smile:

**Comment by fpillet** [comment]
One thing comes up when reading your proposed code -- is the compiler happy about naming a parameter "weak" ? I know we can do it now in Swift 2.2 but you'll need to enclose it in backticks in the function definition

**Comment by IanKeen** [comment]
I'm using this in a 2.2 project 'as-is' doesn't appear to need backticks

**Comment by IanKeen** [comment]
Are you saying that swift 3.0 will need backticks ?

**Comment by fpillet** [comment]
Nope I'm saying that AFAIK Swift 2.2 allows using most keywords as parameter names, but you must backtick them in the function definition (not in the function calls)

**Comment by IanKeen** [comment]
hmm, doesn't appear to be the case for this. I'm using this exact code in a 2.2 project (I've not upgraded to 7.3 yet)

**Comment by fpillet** [comment]
Ok so my bad, I though that we were going to need backticks! all good then.

**Comment by IanKeen** [comment]
:smile: :+1:

**Comment by fpillet** [comment]
Now for this PR -- guess you're going to have fun with the tests suite!

**Comment by IanKeen** [comment]
haha, wish me luck :wink:

**Comment by IanKeen** [comment]
PR #17

**Comment by fpillet** [comment]
Thanks, Ian! Will review this tonite.

**Comment by fpillet** [comment]
Merged  40566f2
